### PR TITLE
Adds Preemptive Memory Limiter for Managing Memory Leaks

### PR DESCRIPTION
**Pull request by ebenoist**
# Adds Preemptive Memory Limiter

Adds a new -m option that when enabled will check the memory limit and usage of the process via the cgroup. When within 10% of the limit, a SIGTERM is sent. This should allow for bette 
memory control for applications that _should_ OOM but grind on their last 10% or so.

## What

Docker's default init script is `tini`, this change extends `tini` to monitor current memory usage via the cgroup. This allows `tini` to interrogate the amount of memory remaining. When 10% of memory is left, `tini` will attempt to SIGTERM the process. It will continue to try until the process comes down.

## Why

We have a number of applications in production with slow memory leaks (e.g. lp-ui, lp-api, cms-api, rql, and bouncer). With all of these applications we've observed the application grow to the limit, but then swap until the application is manually restarted. Docker/ECS should be reaping the instance, but the OOM never comes. 

This gives us a few more knobs to correct this behavior.

## Just Fix the Leaks

We should! We will! This is a bandaid at best, and shouldn't be used as any type of permanent solution. There are a number of better options that we will and should explore to make these errors safer, louder, and more actionable.

### CODE REVIEW

**Review by meshulam — APPROVED** [review]
Makes sense to me

File: .editorconfig [hunk 1]
```diff
@@ -0,0 +1,21 @@
+# EditorConfig is awesome: https://EditorConfig.org
```
Comment (ghost) [inline .editorconfig]: While EditorConfig may indeed be awesome, it has no bearing on this change. This should be part of a separate PR.

File: Dockerfile.alpine [hunk 2]
```diff
@@ -0,0 +1,13 @@
+FROM alpine:3.7
```
Comment (ghost) [inline Dockerfile.alpine]: This is a Dockerfile intended for testing changes introduced in this PR. It should live in `test/` with the rest of the tests.

File: leak.c [hunk 3]
```diff
@@ -0,0 +1,29 @@
+#include <stdio.h>
```
Comment (ghost) [inline leak.c]: This test program should live in the `test/` directory, along with it's respective Dockerfile and the rest of the project's tests.

File: .gitignore [hunk 4]
```diff
@@ -1,3 +1,11 @@
 dist
 sign.key
 .env
+CMakeFiles
+
+CPackConfig.cmake
+CPackSourceConfig.cmake
+Makefile
```
Comment (ghost) [inline .gitignore:8]: I'm assuming these are all generated by the build process, yes?

**Review by ghost — CHANGES_REQUESTED** [review]

File: Dockerfile.alpine (hunk 2 above)
Comment (ebenoist) [reply Dockerfile.alpine]: Test may be a generous word :) -- I'll see if I can make this easier to run/use.

File: .gitignore (hunk 4 above)
Comment (ebenoist) [reply .gitignore:8]: Ya, cmake builds just a couple of files :)

File: .editorconfig (hunk 1 above)
Comment (ebenoist) [reply .editorconfig]: Thats fine, I'll remove it. I just needed to get my vim to work.

**Comment by ebenoist** [comment]
@joshkrueger Move the tests around to be more usable. Let me know if that resolves any issues you have?

**Comment by ghost** [comment]
@ebenoist `test/run_outer_tests.pyc` should not be committed.

File: .editorconfig (hunk 1 above)
Comment (ghost) [reply .editorconfig]: This still needs to be removed from the PR.

**Comment by ebenoist** [comment]
Thanks. Keep having to tweak that gitignore. The original repo didn't have one.

**Comment by ebenoist** [comment]
Should be gtg. I'm going to try adding this to lp-api and see where we're at.
